storage/src/main/java/tech/pegasys/teku/storage/server/ChainStorage.java

```diff
@@ -132,17 +155,67 @@ public SafeFuture<Optional<SignedBeaconBlock>> getEarliestAvailableBlock() {
 
   @Override
   public SafeFuture<Optional<SignedBeaconBlock>> getFinalizedBlockAtSlot(final UInt64 slot) {
-    return SafeFuture.of(() -> database.getFinalizedBlockAtSlot(slot));
+    return SafeFuture.of(() -> database.getFinalizedBlockAtSlot(slot))
+        .thenCompose(this::unblindBlock);
+  }
+
+  private SafeFuture<Optional<SignedBeaconBlock>> unblindBlock(
+      final Optional<SignedBeaconBlock> maybeBlock) {
+    if (maybeBlock.isEmpty()) {
+      return SafeFuture.completedFuture(maybeBlock);
+    }
+    return unblindedBlock(maybeBlock.get()).thenApply(Optional::of);
+  }
+
+  private SafeFuture<SignedBeaconBlock> unblindedBlock(final SignedBeaconBlock block) {
+    if (!block.isBlinded()) {
+      return SafeFuture.completedFuture(block);
+    }
+
+    final Optional<ExecutionPayloadSummary> maybeSummary =
+        block.getMessage().getBody().getOptionalExecutionPayloadSummary();
+    if (maybeSummary.isPresent() && maybeSummary.get().isDefaultPayload()) {
+      // can return without having to fetch anything
+      final SignedBeaconBlock unblinded =
+          block.unblind(
+              spec.atSlot(block.getSlot()).getSchemaDefinitions(),
+              SchemaDefinitionsBellatrix.required(
+                      spec.atSlot(block.getSlot()).getSchemaDefinitions())
+                  .getExecutionPayloadSchema()
+                  .getDefault());
+      return SafeFuture.completedFuture(unblinded);
+    }
+
+    // attempt to fetch payload from storage
+    final Optional<ExecutionPayload> maybePayload =
+        database.getExecutionPayload(block.getRoot(), block.getSlot());
+    if (maybePayload.isPresent()) {
+      return SafeFuture.completedFuture(
+          block.unblind(spec.atSlot(block.getSlot()).getSchemaDefinitions(), maybePayload.get()));
+    }
+
+    if (executionLayerChannel.isEmpty()) {
+      throw new StorageException(
+          String.format(
+              "Needed a payload, but not able to retrieve from storage for block %s (%s)",
+              block.getRoot(), block.getSlot()));
+    }
+
+    throw new StorageException(
+        String.format(
+            "Needed a payload, but not able to retrieve from storage for block %s (%s)",
+            block.getRoot(), block.getSlot()));
```

Comment:
this is where we'd have to do a fetch, currently should never reach here as we're not ever deleting payloads